frappe/model/document.py

```diff
@@ -1250,6 +1253,38 @@ def check_no_back_links_exist(self):
 			check_if_doc_is_linked(self, method="Cancel")
 			check_if_doc_is_dynamically_linked(self, method="Cancel")
 
+	def save_snapshot(self):
+		"""Save snapshot of the document"""
+		if (
+			not self._doc_before_save
+			or self.doctype == "Version"
+			or frappe.flags.in_patch
+			or frappe.flags.in_install
+			or not getattr(self.meta, "allow_document_snapshots", False)
+			or self.docstatus != 0
+		):
+			return
+		old_doc = self._doc_before_save
+		# Remove the name key from the child record, as storing it in snapshot
+		# will remove the child record when restoring the snapshot
+		old_doc = self.delete_name_from_child_table(old_doc)
+		version = frappe.new_doc("Version")
+		version.ref_doctype = self.doctype
+		version.docname = self.name
+		version.complete_snapshot = True
+		version.data = json.dumps(old_doc.as_dict(), default=str)
+		version.insert(ignore_permissions=True)
+
+	def delete_name_from_child_table(self, document: "Document"):
+		for df in document.meta.get_table_fields():
+			child_doc = document.get(df.fieldname, [])
+			if not child_doc:
+				continue
+			for doc in child_doc:
+				doc.delete_key("name")  # Remove the name key from the child record
```

Comment:
Why is this required?

Also `doc.name = None` is better.

Reply:
If a child record contains `name`, it will get removed when the child table is updated.